site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,254 @@
+
+Here's the authorization code for the `app/controllers/application_controller.rb` file:
+
+```ruby
+class ApplicationController < ActionController::API
+  before_action :require_jwt
+
+  def require_jwt
+    token = request.headers["HTTP_AUTHORIZATION"]
+    if !token
+      head :forbidden
+    end
+    if !valid_token(token)
+      head :forbidden
+    end
+  end
+
+  private
+  def valid_token(token)
+    unless token
+      return false
+    end
+
+    token.gsub!('Bearer ','')
+    begin
+      decoded_token = JWT.decode token, Rails.configuration.x.oauth.jwt_secret, true
+      return true
+    rescue JWT::DecodeError
+      Rails.logger.warn "Error decoding the JWT: "+ e.to_s
+    end
+    false
+  end
+end
+```
+
+We look for the JWT in the `AUTHORIZATION` HTTP header. If it doesn't exist, we deny access. If it does, we try to decode it. If it decodes without raising an exception, it is a valid JWT.
+
+## Verify claims
+
+But really, what does valid mean? That's something you define on an application by application basis. The `jwt` gem provides a baseline of default validation: [it checks the `exp` and `nbf` claims](https://github.com/jwt/ruby-jwt/blob/master/lib/jwt/default_options.rb), as well as verifying the signature. 
+
+But for this application, we need to be extra sure. After all, if our messages fell into the wrong hands, who knows what could happen?
+
+So let's verify additional claims are as they should be when we are decoding the JWT. We can do this by providing options to the `JWT.decode` method. Instead of:
+
+```ruby
+# ...
+decoded_token = JWT.decode token, Rails.configuration.x.oauth.jwt_secret, true
+# ...
+```
+
+```ruby
+# ...
+expected_iss = 'fusionauth.io'
+expected_aud = '238d4793-70de-4183-9707-48ed8ecd19d9'
+# ...
+decoded_token = JWT.decode token, Rails.configuration.x.oauth.jwt_secret, true, { verify_iss: true, iss: expected_iss, verify_aud: true, aud: expected_aud, algorithm: 'HS256' }
+# ...
+```
+
+The options at the end of the JWT `decode` method specify which claims we want to verify. If there were private claims that we wanted to check, we'd need to do that as well. Again, while we are guaranteed by the HMAC signature that the contents of the token are exactly what they were when it was created, we aren't guaranteed that the contents will remain unexamined. So add in your needed private claims, but remember JWTs should contain a bare minimum. If needed, consumers can always make additional queries of the identity management server.
```

Comment:
I don't think it's a matter of unexamined as much as a matter of authorizing at this point. Perhaps we say something here that we want to ensure that the issuer is valid and that the JWT is addressed to us (via the aud claim).